homeassistant/components/samsungtv/bridge.py

```diff
@@ -0,0 +1,265 @@
+"""samsungctl and samsungtvws bridge classes."""
+from abc import ABC, abstractmethod
+
+from samsungctl import Remote
+from samsungctl.exceptions import AccessDenied, ConnectionClosed, UnhandledResponse
+from samsungtvws import SamsungTVWS
+from samsungtvws.exceptions import ConnectionFailure
+from websocket import WebSocketException
+
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_ID,
+    CONF_METHOD,
+    CONF_NAME,
+    CONF_PORT,
+    CONF_TIMEOUT,
+    CONF_TOKEN,
+)
+
+from .const import (
+    CONF_DESCRIPTION,
+    LOGGER,
+    METHOD_LEGACY,
+    METHOD_WEBSOCKET,
+    RESULT_AUTH_MISSING,
+    RESULT_NOT_SUCCESSFUL,
+    RESULT_NOT_SUPPORTED,
+    RESULT_SUCCESS,
+    VALUE_CONF_ID,
+    VALUE_CONF_NAME,
+)
+
+
+class SamsungTVBridge(ABC):
+    """The Base Bridge abstract class."""
+
+    @staticmethod
+    def get_bridge(config):
+        """Get Bridge instance."""
+        if config:
+            if config[CONF_METHOD] == METHOD_LEGACY:
+                return SamsungTVLegacyBridge(config)
+            if config[CONF_METHOD] == METHOD_WEBSOCKET:
+                return SamsungTVWSBridge(config)
+        return None
+
+    def __init__(self, config):
+        """Initialize Bridge."""
+        self.port = None
+        self.token = None
+        self.config = config
+        self.method = config[CONF_METHOD]
+        self.host = config[CONF_HOST]
+        self._remote = None
+        self._callback = None
+
+    def register_reauth_callback(self, func):
+        """Register a callback function."""
+        self._callback = func
+
+    @abstractmethod
+    def try_connect(self, port):
+        """Try to connect to the TV."""
+
+    def is_on(self):
+        """Tells if the TV is on."""
+        if self._remote is not None:
+            # Close the current remote connection
+            self._remote.close()
```

Comment:
Shouldn't we use `self.close_remote` here?

Reply:
Currently
`self.close_remote` closes the object returned by the `self._get_remote` method.
`self._get_remote` `if self._remote is None` do creates a new remote object.
`self.close_remote` is used in the `media_player`'s `turn_off `method.

IMHO the `self.close_remote` has to be changed using the above code (no more calls to `self._get_remote`) and then we can use `self.close_remote` as you suggested.
Do you agree?

Reply:
I think it's easier for me to understand if you push your suggested code. You know this logic the best.

Reply:
You're right. done.